File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.cpp
class Solution {
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        if (connections.size() < n - 1) return -1;
        
        vector<vector<int>> adj(n);
        vector<bool> visited(n, false);
        
        for (auto& connection : connections){
            adj[connection[0]].push_back(connection[1]);
            adj[connection[1]].push_back(connection[0]);
        }
        
        int connected = 0;
        for (int node = 0; node < n; ++node){
            if (!visited[node]){
                ++connected;
                visited[node] = true;
                
                queue<int> *q = new queue<int>;
                q -> push(node);
                while (!q -> empty()){
                    int curr = q -> front();
                    q -> pop();
                    for(int neighbor : adj[curr]){
                        if (!visited[neighbor]){
                            visited[neighbor] = true;
                            q -> push(neighbor);
                        }
                    }
                }
            }
        }
        
        return connected - 1;
    }
};

File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.py
class UnionFind:
    def __init__(self, size: int):
        self.parent, self.rank = [], [0] * size
        for i in range(size):
            self.parent.append(i)
            
    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union_set(self, x: int, y: int) -> None:
        xset, yset = self.find(x), self.find(y)
        if self.rank[xset] < self.rank[yset]:
            self.parent[xset] = yset
        elif self.rank[xset] > self.rank[yset]:
            self.parent[yset] = xset
        else:
            self.parent[yset] = xset
            self.rank[xset] += 1
        

class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1
        
        dsu, connected = UnionFind(n), n
        
        for connection in connections:
            if dsu.find(connection[0]) != dsu.find(connection[1]):
                connected -= 1
                dsu.union_set(connection[0], connection[1])
        
        return connected - 1
        
        